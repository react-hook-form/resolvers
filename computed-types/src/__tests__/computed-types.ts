import Schema, { number } from 'computed-types';
import { Resolver, SubmitHandler, useForm } from 'react-hook-form';
import { computedTypesResolver } from '..';
import { fields, invalidData, schema, validData } from './__fixtures__/data';

const shouldUseNativeValidation = false;

describe('computedTypesResolver', () => {
  it('should return values from computedTypesResolver when validation pass', async () => {
    const result = await computedTypesResolver(schema)(validData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toEqual({ errors: {}, values: validData });
  });

  it('should return a single error from computedTypesResolver when validation fails', async () => {
    const result = await computedTypesResolver(schema)(invalidData, undefined, {
      fields,
      shouldUseNativeValidation,
    });

    expect(result).toMatchSnapshot();
  });

  it('should throw any error unrelated to computed-types', async () => {
    const schemaWithCustomError = schema.transform(() => {
      throw Error('custom error');
    });

    const promise = computedTypesResolver(schemaWithCustomError)(
      validData,
      undefined,
      {
        fields,
        shouldUseNativeValidation,
      },
    );

    await expect(promise).rejects.toThrow('custom error');
  });

  /**
   * Type inference tests
   */
  it('should correctly infer the output type from a computedTypes schema', () => {
    const resolver = computedTypesResolver(Schema({ id: number }));

    expectTypeOf(resolver).toEqualTypeOf<
      Resolver<{ id: number }, unknown, { id: number }>
    >();
  });

  it('should correctly infer the output type from a computedTypes schema using a transform', () => {
    const resolver = computedTypesResolver(
      Schema({ id: number.transform((val) => String(val)) }),
    );

    expectTypeOf(resolver).toEqualTypeOf<
      Resolver<{ id: number }, unknown, { id: string }>
    >();
  });

  it('should correctly infer the output type from a computedTypes schema for the handleSubmit function in useForm', () => {
    const schema = Schema({ id: number });

    const form = useForm({
      resolver: computedTypesResolver(schema),
    });

    expectTypeOf(form.watch('id')).toEqualTypeOf<number>();

    expectTypeOf(form.handleSubmit).parameter(0).toEqualTypeOf<
      SubmitHandler<{
        id: number;
      }>
    >();
  });

  it('should correctly infer the output type from a computedTypes schema with a transform for the handleSubmit function in useForm', () => {
    const schema = Schema({ id: number.transform((val) => String(val)) });

    const form = useForm({
      resolver: computedTypesResolver(schema),
    });

    expectTypeOf(form.watch('id')).toEqualTypeOf<number>();

    expectTypeOf(form.handleSubmit).parameter(0).toEqualTypeOf<
      SubmitHandler<{
        id: string;
      }>
    >();
  });
});
