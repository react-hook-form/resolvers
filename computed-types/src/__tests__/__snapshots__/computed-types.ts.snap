// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computedTypesResolver should return a single error from computedTypesResolver when validation fails 1`] = `
Object {
  "errors": Object {
    "birthYear": Object {
      "message": "Expect value to be \\"number\\"",
      "ref": undefined,
      "type": "ValidationError",
    },
    "email": Object {
      "message": "Incorrect email",
      "ref": Object {
        "name": "email",
      },
      "type": "ValidationError",
    },
    "enabled": Object {
      "message": "Expect value to be \\"boolean\\"",
      "ref": undefined,
      "type": "ValidationError",
    },
    "like": Object {
      "message": "Expect value to be \\"string\\"",
      "ref": undefined,
      "type": "ValidationError",
    },
    "password": Object {
      "message": "One uppercase character",
      "ref": Object {
        "name": "password",
      },
      "type": "ValidationError",
    },
    "repeatPassword": Object {
      "message": "Expect value to be \\"string\\"",
      "ref": undefined,
      "type": "ValidationError",
    },
    "tags": Object {
      "message": "Expecting value to be an array",
      "ref": undefined,
      "type": "ValidationError",
    },
    "username": Object {
      "message": "Expect value to be \\"string\\"",
      "ref": Object {
        "name": "username",
      },
      "type": "ValidationError",
    },
  },
  "values": Object {},
}
`;
